File: app/models/invoice.py
"""Modèle Invoice - Facturation des réservations."""
from datetime import date, datetime
from typing import Optional
from sqlalchemy import BigInteger, CheckConstraint, Date, DateTime, ForeignKey, JSON, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.models.base import Base, TimestampMixin, TenantMixin
from app.constants import InvoiceStatus


class Invoice(Base, TimestampMixin, TenantMixin):
    """Facture générée pour une réservation.

    Attributes:
        reservation_id: ID de la réservation (FK unique - one-to-one)
        invoice_number: Numéro de facture unique (ex: "INV-2026-0001")
        issue_date: Date d'émission
        due_date: Date d'échéance (≥ issue_date)
        total_amount: Montant total en CENTIMES
        paid_amount: Montant payé en CENTIMES (≤ total_amount)
        status: Statut (draft, sent, paid, overdue, cancelled)
        payment_method: Moyen de paiement (cash, card, transfer, check)
        payment_date: Date de paiement (nullable)
    """

    __tablename__ = "invoices"

    # Clé primaire
    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)

    # Réservation (relation one-to-many : 1 réservation → N factures selon split)
    reservation_id: Mapped[int] = mapped_column(
        BigInteger,
        ForeignKey("reservations.id", ondelete="CASCADE"),
        nullable=False,
        comment="ID de la réservation (peut avoir 2 factures : advance + balance)"
    )

    # Type de facture (split 40/60)
    invoice_type: Mapped[str] = mapped_column(
        String(10),
        nullable=False,
        default="full",
        comment="Type : full (100%), advance (40%), balance (60%)"
    )

    # Numéro de facture
    invoice_number: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        comment="Numéro de facture unique (INV-2026-0001)"
    )

    # Dates
    issue_date: Mapped[date] = mapped_column(
        Date,
        nullable=False,
        comment="Date d'émission de la facture"
    )

    due_date: Mapped[date] = mapped_column(
        Date,
        nullable=False,
        comment="Date d'échéance de paiement"
    )

    # Montants (en centimes)
    total_amount_cents: Mapped[int] = mapped_column(
        BigInteger,
        nullable=False,
        comment="Montant total en centimes"
    )

    paid_amount_cents: Mapped[int] = mapped_column(
        BigInteger,
        nullable=False,
        default=0,
        comment="Montant payé en centimes"
    )

    # Statut
    status: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default=InvoiceStatus.DRAFT,
        comment="Statut de la facture"
    )

    # Paiement
    payment_method: Mapped[Optional[str]] = mapped_column(
        String(20),
        nullable=True,
        comment="Moyen de paiement (cash, card, transfer, check)"
    )

    payment_date: Mapped[Optional[date]] = mapped_column(
        Date,
        nullable=True,
        comment="Date de paiement"
    )

    # TVA (Option A — taux capturé à la création)
    tva_rate: Mapped[Optional[float]] = mapped_column(
        nullable=True,
        comment="Taux TVA appliqué (ex: 0.20 = 20%) — capturé à la création"
    )

    # Taux d'acompte CGV — capture par facture (peut différer de tenant_settings)
    advance_rate: Mapped[float] = mapped_column(
        nullable=False,
        default=0.40,
        comment="Taux d'acompte CGV appliqué (0.40 = 40%) — capturé à la création facture"
    )

    tva_amount_cents: Mapped[Optional[int]] = mapped_column(
        BigInteger,
        nullable=True,
        comment="Montant TVA en centimes (= total_amount * tva_rate)"
    )

    total_ttc_cents: Mapped[Optional[int]] = mapped_column(
        BigInteger,
        nullable=True,
        comment="Total TTC en centimes (= total_amount + tva_amount_cents)"
    )

    # Timeline audit — timestamps des événements métier clés
    sent_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="Date d'envoi de la facture au client"
    )

    opened_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="Date d'ouverture (suivi email, optionnel)"
    )

    first_reminder_sent_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="Date de la première relance envoyée"
    )

    last_reminder_sent_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="Date de la dernière relance envoyée"
    )

    cancelled_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        comment="Date d'annulation de la facture"
    )

    cancellation_reason: Mapped[Optional[str]] = mapped_column(
        Text,
        nullable=True,
        comment="Raison d'annulation (obligatoire dès qu'on annule).",
    )

    # TVA multi-taux (breakdown par taux)
    tva_breakdown: Mapped[Optional[list]] = mapped_column(
        JSON,
        nullable=True,
        comment=(
            "Détail TVA multi-taux [{'rate':0.20,'base_ht_cents':8000,"
            "'tva_cents':1600,'ttc_cents':9600}]. NULL = taux unique."
        )
    )

    # Relations
    reservation: Mapped["Reservation"] = relationship(
        "Reservation",
        back_populates="invoices"
    )

    charges: Mapped[list["InvoiceCharge"]] = relationship(
        "InvoiceCharge",
        back_populates="invoice",
        cascade="all, delete-orphan"
    )

    payments: Mapped[list["Payment"]] = relationship(
        "Payment",
        back_populates="invoice",
        cascade="all, delete-orphan"
    )

    relances: Mapped[list["Relance"]] = relationship(
        "Relance",
        back_populates="invoice",
        cascade="all, delete-orphan"
    )

    # Contraintes
    __table_args__ = (
        # Numéro de facture unique par tenant
        UniqueConstraint("tenant_id", "invoice_number", name="uq_invoice_tenant_number"),
        # Une réservation ne peut avoir qu'un seul exemplaire de chaque type de facture
        UniqueConstraint("tenant_id", "reservation_id", "invoice_type", name="uq_invoice_tenant_reservation_type"),
        # Dates cohérentes
        CheckConstraint(
            "due_date >= issue_date",
            name="check_invoice_due_after_issue"
        ),
        # Montant payé ≤ total
        CheckConstraint(
            "paid_amount_cents <= total_amount_cents",
            name="check_invoice_paid_lte_total"
        ),
        # Statut valide
        CheckConstraint(
            "status IN ('draft', 'sent', 'paid', 'overdue', 'cancelled')",
            name="check_invoice_status_valid"
        ),
        # Méthode paiement valide
        CheckConstraint(
            "payment_method IS NULL OR payment_method IN ('cash', 'card', 'transfer', 'check')",
            name="check_invoice_payment_method_valid"
        ),
        # Type de facture valide
        CheckConstraint(
            "invoice_type IN ('full', 'advance', 'balance')",
            name="check_invoice_type_valid"
        ),
    )

    @property
    def is_paid(self) -> bool:
        """Facture entièrement payée."""
        return self.paid_amount_cents >= self.total_amount_cents

    @property
    def remaining_amount_cents(self) -> int:
        """Montant restant à payer en centimes."""
        return max(0, self.total_amount_cents - self.paid_amount_cents)

    def __repr__(self) -> str:
        return f"<Invoice(id={self.id}, number='{self.invoice_number}', status='{self.status}')>"

File: frontend/apps/marveline/src/types/invoice.ts
export type InvoiceStatus = 'draft' | 'sent' | 'paid' | 'overdue' | 'cancelled'

export type PaymentMethod = 'cash' | 'card' | 'transfer' | 'check'

export interface InvoiceCharge {
  id: number
  invoice_id: number
  charge_type: 'DAMAGE' | 'LABOR'
  amount_cents: number
  description: string
  hours?: number
  day_type?: 'weekday' | 'weekend' | 'night'
  damage_type_id?: number
  created_at: string
}

export interface InvoiceChargeCreate {
  charge_type: 'DAMAGE' | 'LABOR'
  amount_cents?: number
  hours?: number
  day_type?: 'weekday' | 'weekend' | 'night'
  description: string
}

export interface InvoiceListItem {
  id: number
  tenant_id: number
  reservation_id: number
  invoice_number: string
  issue_date: string
  due_date: string
  status: InvoiceStatus
  total_amount_cents: number
  paid_amount_cents: number
  total_amount_euros: number
  paid_amount_euros: number
  is_paid: boolean
  is_overdue: boolean
  customer_name?: string
  advance_rate: number
  advance_due_date?: string
  sent_at?: string | null
  opened_at?: string | null
  first_reminder_sent_at?: string | null
  last_reminder_sent_at?: string | null
  created_at: string
  updated_at: string
}

export interface ReservationSummary {
  id: number
  reference: string
  status: string
  event_date: string
  customer_name?: string
}

export interface InvoiceDetail extends InvoiceListItem {
  payment_method: PaymentMethod | null
  payment_date: string | null
  reservation: ReservationSummary | null
  remaining_amount_cents: number
  remaining_amount_euros: number
  payment_completion_percentage: number
  charges: InvoiceCharge[]
  tva_rate?: number | null
  tva_amount_cents?: number | null
  tva_amount_euros?: number | null
  total_ttc_cents?: number | null
  total_ttc_euros?: number | null
  tva_breakdown?: TvaBreakdownItem[] | null
  // Timeline audit
  sent_at?: string | null
  opened_at?: string | null
  first_reminder_sent_at?: string | null
  last_reminder_sent_at?: string | null
  cancelled_at?: string | null
}

export interface TvaBreakdownItem {
  rate: number
  base_ht_cents: number
  tva_cents: number
  ttc_cents: number
  base_ht_euros: number
  tva_euros: number
  ttc_euros: number
}

export interface TvaReportResponse {
  month: string
  invoice_count: number
  total_base_ht_cents: number
  total_tva_cents: number
  total_ttc_cents: number
  total_base_ht_euros: number
  total_tva_euros: number
  total_ttc_euros: number
  breakdown_by_rate: TvaBreakdownItem[]
}

export interface AddPaymentRequest {
  amount_cents: number
  payment_method: PaymentMethod
  payment_date: string
}

export interface InvoiceCreateRequest {
  reservation_id: number
  issue_date: string
  due_date: string
  invoice_type?: 'full' | 'advance' | 'balance'
}

export interface InvoiceUpdateRequest {
  issue_date?: string
  due_date?: string
  status?: InvoiceStatus
  total_amount_cents?: number
}

export type InvoiceType = 'invoice' | 'credit_note' | 'damage_invoice'

export interface CreditNote {
  id: number
  original_invoice_id: number
  invoice_number: string
  amount_cents: number
  reason: string
  issue_date: string
  status: 'draft' | 'issued' | 'applied' | 'refunded'
}

export interface InvoiceAuditEntry {
  id: number
  invoice_id: number
  action: string
  user: string
  detail?: string
  created_at: string
}

export interface InvoiceDetailFull extends InvoiceDetail {
  invoice_type: InvoiceType
  tva_rate: number
  subtotal_cents: number
  tva_cents: number
  credit_notes: CreditNote[]
  audit_trail: InvoiceAuditEntry[]
  sent_at?: string
  reminder_count: number
}
